File: .agents/skills/vector-memory/memory/embedder.py
import requests
from typing import List
from .config import OLLAMA_BASE_URL, OLLAMA_EMBED_MODEL, EMBEDDING_DIMENSION

class EmbeddingError(Exception):
    """Raised when embedding generation fails."""
    pass

def get_embedding(text: str) -> List[float]:
    """
    Generate a 768-dimensional embedding vector for given text using local Ollama.
    """
    if not text or not isinstance(text, str) or not text.strip():
        raise ValueError("Text to embed must be a non-empty string.")

    endpoint = f"{OLLAMA_BASE_URL}/api/embeddings"
    payload = {
        "model": OLLAMA_EMBED_MODEL,
        "prompt": text.strip(),
    }

    try:
        response = requests.post(
            endpoint,
            json=payload,
            headers={"Content-Type": "application/json"},
            timeout=30.0,
        )
    except requests.exceptions.ConnectionError as err:
        raise EmbeddingError(
            f"Failed to connect to Ollama at {OLLAMA_BASE_URL}. "
            f"Please ensure Ollama is running (`ollama serve`) and the model is pulled (`ollama pull {OLLAMA_EMBED_MODEL}`)."
        ) from err
    except requests.exceptions.Timeout as err:
        raise EmbeddingError(f"Ollama request timed out after 30s at {OLLAMA_BASE_URL}.") from err
    except Exception as err:
        raise EmbeddingError(f"Unexpected error communicating with Ollama: {err}") from err

    if response.status_code != 200:
        raise EmbeddingError(
            f"Ollama embedding request failed with status {response.status_code}: {response.text}"
        )

    try:
        data = response.json()
    except Exception as err:
        raise EmbeddingError(f"Invalid JSON returned from Ollama: {response.text}") from err

    embedding = data.get("embedding")
    if not isinstance(embedding, list) or len(embedding) == 0:
        raise EmbeddingError(f"Missing or invalid 'embedding' field in Ollama response: {data}")

    if len(embedding) != EMBEDDING_DIMENSION:
        import sys
        print(
            f"[Warning] Expected dimension {EMBEDDING_DIMENSION}, received {len(embedding)}",
            file=sys.stderr,
        )

    return [float(x) for x in embedding]

def get_batch_embeddings(texts: List[str]) -> List[List[float]]:
    """Generate embeddings for a list of texts sequentially."""
    return [get_embedding(t) for t in texts if t and t.strip()]

File: .agents/skills/vector-memory/src/embedder.js
import { CONFIG } from './config.js';

/**
 * Generate a 768-dimensional embedding vector for given text using local Ollama nomic-embed-text.
 * @param {string} text
 * @returns {Promise<number[]>}
 */
export async function getEmbedding(text) {
  if (!text || typeof text !== 'string' || text.trim().length === 0) {
    throw new Error('Text to embed must be a non-empty string.');
  }

  const endpoint = `${CONFIG.ollama.baseUrl}/api/embeddings`;
  const payload = {
    model: CONFIG.ollama.model,
    prompt: text.trim(),
  };

  let response;
  try {
    response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    });
  } catch (err) {
    throw new Error(
      `Failed to connect to Ollama at ${CONFIG.ollama.baseUrl}. Is Ollama running? (${err.message})`
    );
  }

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `Ollama embedding request failed with status ${response.status}: ${errorText}`
    );
  }

  const data = await response.json();
  if (!data.embedding || !Array.isArray(data.embedding)) {
    throw new Error(`Invalid response format from Ollama: ${JSON.stringify(data)}`);
  }

  if (data.embedding.length !== CONFIG.ollama.dimension) {
    console.warn(
      `[Warning] Expected embedding dimension ${CONFIG.ollama.dimension}, received ${data.embedding.length}`
    );
  }

  return data.embedding;
}

/**
 * Generate embeddings for multiple texts sequentially or in parallel batches.
 * @param {string[]} texts
 * @returns {Promise<number[][]>}
 */
export async function getBatchEmbeddings(texts) {
  if (!Array.isArray(texts) || texts.length === 0) {
    return [];
  }

  // Generate embeddings
  const results = [];
  for (const text of texts) {
    const embedding = await getEmbedding(text);
    results.push(embedding);
  }
  return results;
}

export default {
  getEmbedding,
  getBatchEmbeddings,
};
